synapse/rest/client/v2_alpha/register.py

```diff
@@ -428,15 +427,20 @@ async def on_POST(self, request):
                 raise SynapseError(400, "Invalid username")
             desired_username = body["username"]
 
-        appservice = None
-        if self.auth.has_access_token(request):
-            appservice = self.auth.get_appservice_by_req(request)
-
         # fork off as soon as possible for ASes which have completely
         # different registration flows to normal users
 
         # == Application Service Registration ==
-        if appservice:
+        if body.get("type") == AppserviceRegistrationType:
+            if not self.auth.has_access_token(request):
```

Comment:
Maybe we should check for this access token regardless of type, but throw if the type is wrong to help people diagnose their broken appservices?

Reply:
Have done so in 8bd7eb6